readthedocs/notifications/querysets.py

```diff
@@ -22,13 +24,63 @@ def add(self, *args, **kwargs):
             attached_to_content_type=content_type,
             attached_to_id=attached_to.id,
             message_id=message_id,
+            # Update only ``READ`` and ``UNREAD`` notifications because we want
+            # to keep track of ``DISMISSED`` and ``CANCELLED`` ones.
+            state__in=(UNREAD, READ),
         ).first()
 
         if notification:
             self.filter(pk=notification.pk).update(
-                *args, modified=timezone.now(), **kwargs
+                *args,
+                modified=timezone.now(),
+                state=UNREAD,
+                **kwargs,
             )
             notification.refresh_from_db()
             return notification
 
         return super().create(*args, attached_to=attached_to, **kwargs)
+
+    def cancel(self, message_id, attached_to):
+        """
+        Cancel an on-going notification because the underlying state has changed.
+
+        When a notification is not valid anymore because the user has made the
+        required action (e.g. paid an unpaid subscription) we use this method
+        to mark those notifications as ``CANCELLED``.
+
+        It only cancels notifications that are ``UNREAD`` or ``READ``.
```

Comment:
This isn't super clear.. what other states are there that we should care about? 🤔